### PR TITLE
Dynamo x autograd.Function supports non-{Tensor, symnode, constant} inputs

### LINKED ISSUE #118395: Dynamo x autograd.Function: graph breaks on non-{Tensor, symnode, constant} inputs
```
import torch

class MulY(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x, ys):
        ctx.save_for_backward(ys[0])
        return x * ys[0]

    @staticmethod
    def backward(ctx, grad_out):
        ys, = ctx.saved_tensors 
        return grad_out * ys

@torch.compile(backend="eager", fullgraph=True)
def f(x, ys):
    return MulY.apply(x, ys)

x = torch.tensor(2., requires_grad=True)
ys = [torch.tensor(2.)]
expected = MulY.apply(x, ys)
out = f(x, ys)
assert torch.allclose(out, expected)
```

We should be able to support this with fullgraph=True
We should be able to do something like manually_set_inputs=False for autograd.Function HOP tracing.

cc @ezyang @gchanan @kadeng @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #118395

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

**Comment by yanboliang** [comment]
@zou3519 I posted this PR to get some early feedbacks, I need to update documents later if this is the proper solution.

File: torch/_functorch/autograd_function.py [hunk 1]
```diff
@@ -686,21 +686,35 @@ class AutogradFunctionApply(HigherOrderOperator):
...
+        # lifted freevars should not be args of ApplyTemplate.apply
+        # since we don't need to calculate the gradients of them.
+        new_fwd_args = fwd_args[:length_of_tensor_args]
 
         class ApplyTemplate(torch.autograd.Function):
             @staticmethod
             def forward(ctx, *args):
                 nonlocal saved_values
-                output, saved_values = fwd(None, *args)
+                output, saved_values = fwd(None, *fwd_args)
```
Comment (zou3519) [inline torch/_functorch/autograd_function.py:702]: I'm not sure this is going to work (but if it does, then go for it). I'm worried that autograd.Function might require that the `args` that get passed in actually be used, so something like:
```
class ApplyTemplate
  @staticmethod
  def forward(ctx, *args):
        actual_args = (*args, *additional_args)
        output, saved_values = fwd(None, *actual_args)
```

File: torch/_functorch/autograd_function.py [hunk 2]
```diff
@@ -686,21 +686,35 @@ class AutogradFunctionApply(HigherOrderOperator):
...
                 return output
 
             @staticmethod
             def backward(ctx, *grad):
-                return bwd(None, *grad, *saved_values)
-
-        return ApplyTemplate.apply(*fwd_args)
+                output = bwd(None, *grad, *saved_values)
+                # The output includes the gradients of all original args, which can be None for non-Tensor args.
+                # We need to filter out the None values.
```
Comment (zou3519) [inline torch/_functorch/autograd_function.py:709]: The gradient can also be None for Tensor args, so we need to keep track of which inputs are not actually Tensors

File: torch/_functorch/autograd_function.py [hunk 3]
```diff
@@ -686,21 +686,35 @@ class AutogradFunctionApply(HigherOrderOperator):
...
+                if isinstance(output, tuple):
+                    return tuple(
+                        [x for x, mask in zip(output, args_tensor_mask) if mask]
+                    )
+                else:
+                    return output
```
Comment (zou3519) [inline torch/_functorch/autograd_function.py:715]: Btw, @Chillee's suggestion at the sync might have been to do the rewrite of the backward graph in Dynamo:
1) speculate subgraph tells us which arguments we actually used. In the weird case, it would say something like `(from_arg(0), from_arg(2), "freevar")`
2) Based on this, we would get Dynamo to rewrite the `bwd` graph to drop output 1:
```
# rewrite the following
def bwd(ctx, grad, saved_vars):
    return <Tensor>, None, <Tensor>

# into this:
def bwd(ctx, grad, saved_vars):
    return <Tensor>, <Tensor>
```
All of this would happen in Dynamo (and outside of the autograd_function HOP (this file))

**Review by zou3519 — COMMENTED** [review]
approach LGTM, we may want some more testing around edge cases (e.g. capture more than one variable, the inputs are in a different order than they are used, returning None as the gradient for a Tensor, TensorList input)

File: torch/_functorch/autograd_function.py (hunk 2 above)
Comment (yanboliang) [reply torch/_functorch/autograd_function.py:709]: Yes, I use ```args_tensor_mask``` to track which inputs are Tensors, probably I need to make the documents here more accurate.

File: torch/_functorch/autograd_function.py (hunk 1 above)
Comment (yanboliang) [reply torch/_functorch/autograd_function.py:702]: I think it's the same as what I did: ```fwd_args``` = ```args``` + lifted freevars(which is ```additional_args``` in your example).

File: torch/_functorch/autograd_function.py (hunk 3 above)
Comment (yanboliang) [reply torch/_functorch/autograd_function.py:715]: Oh, I think we just need to move the rewrite from this file to _dynamo/variables/higher_order_ops.py, which I think it's doable. I'm curious why that option is better than the current one?

**Comment by yanboliang** [comment]
> approach LGTM, we may want some more testing around edge cases (e.g. capture more than one variable, the inputs are in a different order than they are used, returning None as the gradient for a Tensor, TensorList input)

SG, I'll add more comprehensive tests.

File: torch/_functorch/autograd_function.py (hunk 3 above)
Comment (zou3519) [reply torch/_functorch/autograd_function.py]: tbh it's probably easier to do the backward rewrite in this file. I think Horace's point is more of (1): speculate subgraph should tell us which arguments we actually used (this is useful for other HOPs)

File: test/dynamo/test_repros.py [hunk 4]
```diff
@@ -1160,11 +1124,11 @@ def test_reformer_train(self):
...
         if torch._dynamo.config.assume_static_by_default:
-            self.assertExpectedInline(cnt.frame_count, """3""")
-            self.assertExpectedInline(cnt.op_count, """10""")
+            self.assertExpectedInline(cnt.frame_count, """1""")
+            self.assertExpectedInline(cnt.op_count, """5""")
         else:
-            self.assertExpectedInline(cnt.frame_count, """3""")
-            self.assertExpectedInline(cnt.op_count, """10""")
+            self.assertExpectedInline(cnt.frame_count, """1""")
+            self.assertExpectedInline(cnt.op_count, """5""")
```
Comment (yanboliang) [inline test/dynamo/test_repros.py:1131]: The reason of changing this unit test: The original backward implementation is wrong. It was not tested because we don't trace the backward function before this change. This was copied from modeling_reformer.py in huggingface, I found it's not easy to copy everything updated from there, so I made a simplification.

File: test/dynamo/test_autograd_function.py [hunk 5]
```diff
@@ -467,32 +469,116 @@ def fn(a, b):
...
+            b: torch.Tensor
+            c: torch.Tensor
 
         class Foo(torch.autograd.Function):
             @staticmethod
-            def forward(ctx, x):
-                return x.clone(), (x + delta).clone()
+            def forward(ctx, x: torch.Tensor, weird: Weird, z: torch.Tensor):
+                ctx.save_for_backward(weird.b, weird.c)
+                return x.clone() * weird.b * weird.c
```
Comment (zou3519) [inline test/dynamo/test_autograd_function.py:483]: Can you change this to `weird.b * weird.c * x.clone()`? (To test that the order things are accessed doesn't matter)

File: test/dynamo/test_autograd_function.py [hunk 6]
```diff
@@ -467,32 +469,116 @@ def fn(a, b):
...
 
-        self.assertEqual(result, Foo.apply(x))
-        self.assertEqual(cnt.frame_count, 1)
-        self.assertEqual(
-            list(torch._dynamo.utils.counters["graph_break"].values()), [1]
-        )
+        result = f(x, weird, z)
+        result.sum().backward()
+
+        self.assertEqual(result, Foo.apply(x, weird, z))
```
Comment (zou3519) [inline test/dynamo/test_autograd_function.py:501]: we should test backend="aot_eager" and check that the gradients are as we expect. So we should check x.grad and z.grad

File: test/dynamo/test_autograd_function.py [hunk 7]
```diff
@@ -467,32 +469,116 @@ def fn(a, b):
...
+        def f(x, tl):
+            return Foo.apply(x, tl)
+
+        x = torch.tensor(2.0, requires_grad=True)
+        tl = [torch.tensor(3.0), torch.tensor(4.0)]
+
+        result = f(x, tl)
+        result.sum().backward()
+
+        self.assertEqual(result, Foo.apply(x, tl))
```
Comment (zou3519) [inline test/dynamo/test_autograd_function.py:525]: Ditto here: should test aot_eager and that x.grad is what we expect

File: torch/_dynamo/variables/higher_order_ops.py [hunk 8]
```diff
@@ -1575,6 +1603,59 @@ def is_strict_for(v: VariableTracker):
...
+        # This is used to determine if a FX node's argument should be an argument of
+        # ApplyTemplate.forward and if we should skip the output from ApplyTemplate.backward
+        # at torch._functorch.autograd_function.AutogradFunctionApply.
+        args_tensor_mask = [False] * len(args)
+        for i, arg in enumerate(args):
+            if isinstance(arg, (variables.TensorVariable, variables.SymNodeVariable)):
+                filtered_args.append(arg)
+                args_tensor_mask[i] = True
+
+        # Rewrite the output of bwd_graph to remove the grad output for the non-Tensor args.
```
Comment (zou3519) [inline torch/_dynamo/variables/higher_order_ops.py:1621]: is it possible to assert that they are None?

File: test/dynamo/test_autograd_function.py [hunk 9]
```diff
@@ -467,32 +469,116 @@ def fn(a, b):
         self.assertEqual(res, MyMM.apply(a, a))
         self.assertEqual(cnt.frame_count, 1)
 
-    def test_graph_break_if_lifted_free_variable(self):
-        torch._dynamo.utils.counters.clear()
-        cnt = torch._dynamo.testing.CompileCounter()
-        delta = torch.randn(3)
+    def test_user_defined_object_as_input(self):
```
Comment (zou3519) [inline test/dynamo/test_autograd_function.py:472]: Can you add an assertExpectedInline for the code generated by Dynamo for this case?

File: torch/_functorch/autograd_function.py (hunk 1 above)
Comment (zou3519) [reply torch/_functorch/autograd_function.py:702]: Not exactly -- the code in this PR looks like:
```
class ApplyTemplate(torch.autograd.Function):
    @staticmethod
     def forward(ctx, *args):
         nonlocal saved_values
         output, saved_values = fwd(None, *fwd_args)
         return output
```
note that `args` isn't actually passed directly to fwd: we pass in `fwd_args` to `fwd`. 

`fwd_args` contains the tensors in `args`, but... I'm not sure if that works. Again, if the tests pass then I'm happy with it.

**Review by zou3519 — COMMENTED** [review]
Just have some testing comments:
- we should test backend="aot_eager" and also check that the gradients (e.g. x.grad) are what we expect them to be.
- we should add some self.assertExpectedInline tests for the graph being generated by Dynamo so it's clear what is going on.

The code looks reasonable to me as long as the tests pass

File: test/dynamo/test_autograd_function.py (hunk 6 above)
Comment (zou3519) [reply test/dynamo/test_autograd_function.py:501]: For example, x.grad should be 2.5 * 3.5; z.grad should be 2.

Maybe also set requires_grad=True on weird.b and then check that it doesn't have a grad.

File: torch/_dynamo/variables/higher_order_ops.py (hunk 8 above)
Comment (yanboliang) [reply torch/_dynamo/variables/higher_order_ops.py:1621]: Added assertion at L1651

**Review by zou3519 — APPROVED** [review]
looks great!
